mbf_abstract_nav/src/abstract_controller_execution.cpp

```diff
@@ -185,6 +186,51 @@ void AbstractControllerExecution::setVelocityCmd(const geometry_msgs::TwistStamp
   // TODO so there should be no loss of information in the feedback stream
 }
 
+bool AbstractControllerExecution::checkVelocityIgnore(const bool& robot_stopped,
+                                                      const geometry_msgs::TwistStamped& cmd_velocity)
+{
+  bool ignored_exceed_tolerance = false;
+
+  if (robot_stopped)
+  {
+    // linear total and angular
+    double cmd_linear;
+    double cmd_angular;
+
+    // compute linear velocity magnitude
+    cmd_linear = hypot(cmd_velocity.twist.linear.x, cmd_velocity.twist.linear.y);
+    // compute angular velocity magnitude
+    cmd_angular = fabs(cmd_velocity.twist.angular.z);
+
+    if (cmd_linear > 0.01 || cmd_angular > 0.01)  // if cmd_vel is non-zero
+    {
+      // output robot ignored velocity command warning message
+      double ignored_duration = (ros::Time::now() - first_ignored_time_).toSec();
+      ROS_WARN_THROTTLE(1, "Robot ignores velocity command for %.2f seconds", ignored_duration);
+
+      if (ignored_duration > robot_ignore_check_tolerance_)
+      {
+        // the robot is ignoring the velocity command exceed the threshold time
+        ROS_ERROR("Robot is disabled and the time it ignored velocity command exceeded the tolerance: %.2f seconds",
+                  robot_ignore_check_tolerance_);
+        ignored_exceed_tolerance = true;
+      }
+    }
+    else  // robot stopped but velocity command is also close to zero
+    {
+      ROS_DEBUG("Robot ignore velocity command check is ok");
+      first_ignored_time_ = ros::Time::now();
+    }
+  }
+  else  // robot is moving, not need to check
+  {
+    ROS_DEBUG("Robot ignore velocity command check is ok");
+    first_ignored_time_ = ros::Time::now();
+  }
+
+  return ignored_exceed_tolerance;
```

Comment:
S: Avoid doing nested if statements. It's more readable to do something like this:

```python
if is_robot_moving or is_cmd_vel_zero:
    return false

# do main logic here ...
```